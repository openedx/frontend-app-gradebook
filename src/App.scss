
@import "~@edx/paragon/scss/edx/theme.scss";

$fa-font-path: "~font-awesome/fonts";
@import "~font-awesome/scss/font-awesome";

$input-focus-box-shadow: $input-box-shadow; // hack to get upgrade to paragon 4.0.0 to work
//@import "~@edx/paragon/src/SearchField/SearchField";
//@import "~@edx/paragon/src/Collapsible/Collapsible";

@import "~@edx/frontend-component-footer/src/lib/scss/site-footer";

@import "./components/Gradebook/gradebook";
@import "./components/Drawer/Drawer";

