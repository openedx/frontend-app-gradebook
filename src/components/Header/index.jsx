import React from 'react';
import SiteHeader from './SiteHeader';
import { Hyperlink } from '@edx/paragon';

const Header = () => {
  return (
    <SiteHeader 
      menuItems={MENU_ITEMS}
      desktopMenuItems={DESKTOP_MENU_ITEMS}
    />
  );
}

export default Header;



const MENU_ITEMS = [
  {
    content: "Courses",
    name: "courses",
    destination: "#",
    submenu: (
      <div>
        <div>
          <h3>Courses by subject</h3>
        </div>
        <ul>
          <li><Hyperlink content="Computer Science" destination="#" /></li>
          <li><Hyperlink content="Language" destination="#" /></li>
          <li><Hyperlink content="Data & Statistics" destination="#" /></li>
          <li><Hyperlink content="Business & Management" destination="#" /></li>
          <li><Hyperlink content="Engineering" destination="#" /></li>
          <li><Hyperlink content="Humanities" destination="#" /></li>
          <li><Hyperlink content="View all courses by subject" destination="#" /></li>
        </ul>
      </div>
    )
  },
  {
    content: "Programs",
    name: "programs",
    destination: "#",
    submenu: (
      <div>
        <h3>Programs & degrees</h3>
        <p>Pathways for your advancement</p>
        <ul>
          <li>
            <Hyperlink content="MicroMasters Program" destination="#" />
            <p>Graduate-level, for career advancement or a degree path</p>
          </li>

          <li>
            <Hyperlink content="Professional Certificate" destination="#" />
            <p>From employers or universities to build today's in-demand skills</p>
          </li>

          <li>
            <Hyperlink content="Online Master's Degree" destination="#" />
            <p>Top-ranked programs, affordable, and fully online</p>
          </li>

          <li>
            <Hyperlink content="Global Freshman Academy" destination="#" />
            <p>Freshman year courses for university credit from ASU</p>
          </li>

          <li>
            <Hyperlink content="XSeries" destination="#" />
            <p>Series of courses for a deep understanding of a topic</p>
          </li>
        </ul>
      </div>
    )
  },
  {
    content: "Schools",
    name: "schools",
    destination: "#"
  },
  {
    content: "Enterprise",
    name: "enterprise",
    destination: "#"
  }
]

const DESKTOP_MENU_ITEMS = [
  {
    content: "Courses",
    name: "courses",
    destination: "#",
    submenu: (
      <div className="main-menu-courses">
        <div>
          <h3>Courses by subject</h3>
        </div>
        <ul>
          <li><Hyperlink content="Architecture" destination="#" /></li>
          <li><Hyperlink content="Art & Culture" destination="#" /></li>
          <li><Hyperlink content="Biology & Life Sciences" destination="#" /></li>
          <li><Hyperlink content="Business & Management" destination="#" /></li>
          <li><Hyperlink content="Chemistry" destination="#" /></li>
          <li><Hyperlink content="Communication" destination="#" /></li>
          <li><Hyperlink content="Computer Science" destination="#" /></li>
          <li><Hyperlink content="Data Analysis & Statistics" destination="#" /></li>
          <li><Hyperlink content="Design" destination="#" /></li>
          <li><Hyperlink content="Economics & Finance" destination="#" /></li>
          <li><Hyperlink content="Education & Teacher Training" destination="#" /></li>
          <li><Hyperlink content="Electronics" destination="#" /></li>
          <li><Hyperlink content="Energy & Earth Sciences" destination="#" /></li>
          <li><Hyperlink content="Engineering" destination="#" /></li>
          <li><Hyperlink content="Environmental Studies" destination="#" /></li>
          <li><Hyperlink content="Ethics" destination="#" /></li>
          <li><Hyperlink content="Food & Nutrition" destination="#" /></li>
          <li><Hyperlink content="Health & Safety" destination="#" /></li>
          <li><Hyperlink content="History" destination="#" /></li>
          <li><Hyperlink content="Humanities" destination="#" /></li>
          <li><Hyperlink content="Language" destination="#" /></li>
          <li><Hyperlink content="Law" destination="#" /></li>
          <li><Hyperlink content="Literature" destination="#" /></li>
          <li><Hyperlink content="Math" destination="#" /></li>
          <li><Hyperlink content="Medicine" destination="#" /></li>
          <li><Hyperlink content="Music" destination="#" /></li>
          <li><Hyperlink content="Philanthropy" destination="#" /></li>
          <li><Hyperlink content="Philosophy & Ethics" destination="#" /></li>
          <li><Hyperlink content="Physics" destination="#" /></li>
          <li><Hyperlink content="Science" destination="#" /></li>
          <li><Hyperlink content="Social Sciences" destination="#" /></li>
          <li><Hyperlink content="All Subjects Â»" destination="#" /></li>


        </ul>
      </div>
    )
  },
  {
    content: "Programs",
    name: "programs",
    destination: "#",
    submenu: (
      <div>
        <h3>Programs & degrees</h3>
        <p>Pathways for your advancement</p>
        <ul>
          <li>
            <Hyperlink content="MicroMasters Program" destination="#" />
            <p>Graduate-level, for career advancement or a degree path</p>
          </li>

          <li>
            <Hyperlink content="Professional Certificate" destination="#" />
            <p>From employers or universities to build today's in-demand skills</p>
          </li>

          <li>
            <Hyperlink content="Online Master's Degree" destination="#" />
            <p>Top-ranked programs, affordable, and fully online</p>
          </li>

          <li>
            <Hyperlink content="Global Freshman Academy" destination="#" />
            <p>Freshman year courses for university credit from ASU</p>
          </li>

          <li>
            <Hyperlink content="XSeries" destination="#" />
            <p>Series of courses for a deep understanding of a topic</p>
          </li>
        </ul>
      </div>
    )
  },
  {
    content: "Schools",
    name: "schools",
    destination: "#"
  },
  {
    content: "Enterprise",
    name: "enterprise",
    destination: "#"
  }
]