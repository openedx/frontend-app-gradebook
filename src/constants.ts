export const appId = 'org.openedx.frontend.app.gradebook';
