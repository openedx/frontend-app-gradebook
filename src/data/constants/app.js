import { StrictDict } from 'utils';
import { getConfig } from '@edx/frontend-platform';

export const routePath = `${getConfig().PUBLIC_PATH}:courseId`;

export const views = StrictDict({
  grades: 'grades',
  bulkManagementHistory: 'bulkManagementHistory',
});

export const modalFieldKeys = StrictDict({
  adjustedGradePossible: 'adjustedGradePossible',
  adjustedGradeValue: 'adjustedGradeValue',
  assignmentName: 'assignmentName',
  reasonForChange: 'reasonForChange',
  todaysDate: 'todaysDate',
  updateModuleId: 'updateModuleId',
  updateUserId: 'updateUserId',
  updateUserName: 'updateUserName',
  open: 'open',
});

export const localFilterKeys = StrictDict({
  assignmentGradeMax: 'assignmentGradeMax',
  assignmentGradeMin: 'assignmentGradeMin',
  courseGradeMax: 'courseGradeMax',
  courseGradeMin: 'courseGradeMin',
});

/**
 * column configuration for bulk management tab's data table
 */
export const bulkManagementColumns = [
  {
    accessor: 'filename',
    Header: 'Gradebook',
    columnSortable: false,
    width: 'col-5',
  },
  {
    accessor: 'resultsSummary',
    Header: 'Download Summary',
    columnSortable: false,
    width: 'col',
  },
  {
    accessor: 'user',
    Header: 'Who',
    columnSortable: false,
    width: 'col-1',
  },
  {
    accessor: 'timeUploaded',
    Header: 'When',
    columnSortable: false,
    width: 'col',
  },
];

export const gradeOverrideHistoryColumns = StrictDict({
  adjustedGrade: 'adjustedGrade',
  date: 'date',
  grader: 'grader',
  reason: 'reason',
});

/**
 * Display strings for various app components.
 * Note: this is a temporary storage location for these strings, before we put them in
 * messages files for translation.
 */
export const messages = StrictDict({
  BulkManagementTab: StrictDict({
    csvUploadLabel: 'Upload Grade CSV',
    heading: 'Use this feature by downloading a CSV for bulk management, overriding grades locally, and coming back here to upload.',
    importBtnText: 'Import Grades',
    successDialog: 'CSV processing. File uploads may take several minutes to complete.',
    hints: [
      'Results appear in the table below.',
      'Grade processing may take a few seconds.',
    ],
  }),
});
