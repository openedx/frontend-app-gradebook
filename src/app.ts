import { App } from '@openedx/frontend-base';
import { appId } from './constants';
import routes from './routes';
import providers from './providers';
import messages from './i18n';
import slots from './slots';

const app: App = {
  appId,
  routes,
  providers,
  messages,
  slots,
  config: {
    LEARNING_BASE_URL: 'http://apps.local.openedx.io:2000',
    ENABLE_PROGRAMS: false,
    ECOMMERCE_BASE_URL: '',
    ORDER_HISTORY_URL: '',
    SUPPORT_URL: '',
  }
};

export default app;
