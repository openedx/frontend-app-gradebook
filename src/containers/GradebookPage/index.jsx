/* eslint-disable import/no-named-as-default */
import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import queryString from 'query-string';

import selectors from 'data/selectors';
import thunkActions from 'data/thunkActions';
import { views } from 'data/constants/app';

import WithSidebar from 'components/WithSidebar';
import GradebookHeader from 'components/GradebookHeader';
import GradesView from 'components/GradesView';
import GradebookFilters from 'components/GradebookFilters';
import BulkManagementHistoryView from 'components/BulkManagementHistoryView';

/**
 * <GradebookPage />
 * Top-level view for the Gradebook MFE.
 * Organizes a header and a pair of views (Grades and BulkManagement) with a toggle-able
 * filter sidebar.
 */
export class GradebookPage extends React.Component {
  constructor(props) {
    super(props);
    this.updateQueryParams = this.updateQueryParams.bind(this);
  }

  componentDidMount() {
    const urlQuery = queryString.parse(this.props.location.search);
    this.props.initializeApp(this.props.match.params.courseId, urlQuery);
  }

  updateQueryParams(queryParams) {
    const parsed = queryString.parse(this.props.location.search);
    Object.keys(queryParams).forEach((key) => {
      if (queryParams[key]) {
        parsed[key] = queryParams[key];
      } else {
        delete parsed[key];
      }
    });
    this.props.history.push(`?${queryString.stringify(parsed)}`);
  }

  render() {
    return (
      <WithSidebar
        sidebar={<GradebookFilters updateQueryParams={this.updateQueryParams} />}
      >
        <div className="px-3 gradebook-content">
          <GradebookHeader />
          {(this.props.activeView === views.bulkManagementHistory
            ? <BulkManagementHistoryView />
            : <GradesView updateQueryParams={this.updateQueryParams} />
          )}
        </div>
      </WithSidebar>
    );
  }
}
GradebookPage.defaultProps = {
  location: { search: '' },
};
GradebookPage.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func,
  }).isRequired,
  location: PropTypes.shape({ search: PropTypes.string }),
  match: PropTypes.shape({
    params: PropTypes.shape({
      courseId: PropTypes.string,
    }),
  }).isRequired,
  // redux
  activeView: PropTypes.string.isRequired,
  initializeApp: PropTypes.func.isRequired,
};

export const mapStateToProps = (state) => ({
  activeView: selectors.app.activeView(state),
});

export const mapDispatchToProps = {
  initializeApp: thunkActions.app.initialize,
};

export default connect(mapStateToProps, mapDispatchToProps)(GradebookPage);
