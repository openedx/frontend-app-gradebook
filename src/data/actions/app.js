import { StrictDict } from 'utils';
import { modalFieldKeys, localFilterKeys } from 'data/constants/app';
import { createActionFactory } from './utils';

export const dataKey = 'app';
const createAction = createActionFactory(dataKey);

const closeModal = createAction('closeModal');
/**
 * setCourseId(courseId)
 * loads courseID into local app state
 * @param {string} courseId - course ID from app context
 */
const setCourseId = createAction('setCourseId');

const filterMenu = StrictDict({
  endTransition: createAction('filterMenu/endTransition'),
  startTransition: createAction('filterMenu/startTransition'),
  toggle: createAction('filterMenu/toggle'),
});

/**
  * setModalStateFromTable({ userEntry, subsection })
  * sets modalState
  */
const setModalStateFromTable = createAction('setModalStateFromTable');

/**
 * setSearchValue(searchValue)
 * sets searchValue in local state
 * @param {string} searchValue - search input text
 */
const setSearchValue = createAction('setSearchValue');

const filterReducer = (payload) => (obj, key) => (
  (payload[key] !== undefined) ? { ...obj, [key]: payload[key] } : obj
);

/**
 * setLocalFilter(filterObject)
 * sets a number of localFilterKey fields on localFilters
 * @param {object} filterObject - an object of filter values
 */
const setLocalFilter = createAction('setLocalFilter', (filterObject) => ({
  payload: Object.keys(localFilterKeys).reduce(filterReducer(filterObject), {}),
}));

/**
 * setModalState(modalState)
 * sets a number of modalFieldKey fields on modalState
 * @param {object} modalState - an object of modal state values
 */
const setModalState = createAction('setModalState', (modalState) => ({
  payload: Object.keys(modalFieldKeys).reduce(filterReducer(modalState), {}),
}));

/**
 * setShowImportSuccessToast(shouldShow)
 * Set whether or not to show the Import Grades success toast
 * @param {bool} sholdShow - should show the toast?
 */
const setShowImportSuccessToast = createAction('setShowImportSuccessToast');

/**
 * setView(viewId)
 * sets the UI to display the tab indcated by the passed view id
 * @param {string} viewId - view id as set in app constants
 */
const setView = createAction('setView');

export default StrictDict({
  closeModal,
  filterMenu,
  setCourseId,
  setLocalFilter,
  setModalState,
  setModalStateFromTable,
  setSearchValue,
  setShowImportSuccessToast,
  setView,
});
