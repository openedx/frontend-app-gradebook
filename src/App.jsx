import React from 'react';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';

import { AppProvider } from '@edx/frontend-platform/react';

import Footer from '@edx/frontend-component-footer';
import Header from '@edx/frontend-component-header';

import { routePath } from 'data/constants/app';
import store from 'data/store';
import GradebookPage from 'containers/GradebookPage';
import './App.scss';
import Head from './head/Head';

const App = () => (
  <AppProvider store={store}>
    <Head />
    <Router>
      <div>
        <Header />
        <main>
          <Switch>
            <Route
              exact
              path={routePath}
              component={GradebookPage}
            />
          </Switch>
        </main>
        <Footer logo={process.env.LOGO_POWERED_BY_OPEN_EDX_URL_SVG} />
      </div>
    </Router>
  </AppProvider>
);

export default App;
