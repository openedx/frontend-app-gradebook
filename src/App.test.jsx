import React from 'react';
import { shallow } from 'enzyme';

import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import { AppProvider } from '@edx/frontend-platform/react';

import Footer from '@edx/frontend-component-footer';
import Header from '@edx/frontend-component-header';

import { routePath } from 'data/constants/app';
import store from 'data/store';
import GradebookPage from 'containers/GradebookPage';

import App from './App';
import Head from './head/Head';

jest.mock('react-router-dom', () => ({
  BrowserRouter: () => 'BrowserRouter',
  Route: () => 'Route',
  Switch: () => 'Switch',
}));
jest.mock('@edx/frontend-platform/react', () => ({
  AppProvider: () => 'AppProvider',
}));
jest.mock('data/constants/app', () => ({
  routePath: '/:courseId',
}));
jest.mock('@edx/frontend-component-footer', () => 'Footer');
jest.mock('data/store', () => 'testStore');
jest.mock('containers/GradebookPage', () => 'GradebookPage');
jest.mock('@edx/frontend-component-header', () => 'Header');

const logo = 'fakeLogo.png';
let el;
let router;

describe('App router component', () => {
  test('snapshot', () => {
    expect(shallow(<App />)).toMatchSnapshot();
  });
  describe('component', () => {
    beforeEach(() => {
      process.env.LOGO_POWERED_BY_OPEN_EDX_URL_SVG = logo;
      el = shallow(<App />);
      router = el.childAt(1);
    });
    describe('AppProvider', () => {
      test('AppProvider is the parent component, passed the redux store props', () => {
        expect(el.type()).toBe(AppProvider);
        expect(el.props().store).toEqual(store);
      });
    });
    describe('Head', () => {
      test('first child of AppProvider', () => {
        expect(el.childAt(0).type()).toBe(Head);
      });
    });
    describe('Router', () => {
      test('second child of AppProvider', () => {
        expect(router.type()).toBe(Router);
      });
      test('Header is above/outside-of the routing', () => {
        expect(router.childAt(0).childAt(0).type()).toBe(Header);
        expect(router.childAt(0).childAt(1).type()).toBe('main');
      });
      test('Routing - GradebookPage is only route', () => {
        expect(router.find('main')).toEqual(shallow(
          <main>
            <Switch>
              <Route exact path={routePath} component={GradebookPage} />
            </Switch>
          </main>,
        ));
      });
    });
    test('Footer logo drawn from env variable', () => {
      expect(router.find(Footer).props().logo).toEqual(logo);
    });
  });
});
