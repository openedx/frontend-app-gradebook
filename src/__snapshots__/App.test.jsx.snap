// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App router component snapshot 1`] = `
<AppProvider
  store="testStore"
>
  <injectIntl(ShimmedIntlComponent) />
  <BrowserRouter>
    <div>
      <Header />
      <main>
        <Switch>
          <Route
            component="GradebookPage"
            exact={true}
            path="/:courseId"
          />
        </Switch>
      </main>
      <Footer />
    </div>
  </BrowserRouter>
</AppProvider>
`;
